### PR TITLE
gNOI-2.1 packet link qualification deviations

### PR DESCRIPTION
**Pull request by arvbaska1**
Adding deviations for gNOI-2.1 packet link qualification test

### CODE REVIEW

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go [hunk 1]
```diff
@@ -422,11 +439,16 @@ func TestLinkQualification(t *testing.T) {
 	// The packet counters between Generator and Reflector mismatch tolerance level in percentage
 	var tolerance float64 = 0.0001
 
-	if ((math.Abs(float64(generatorPktsSent)-float64(reflectorPktsRxed)))/(float64(generatorPktsSent)+float64(reflectorPktsRxed)+tolerance))*200.00 > tolerance {
-		t.Errorf("The difference between packets received count at Reflector and packets sent count at Generator is greater than %0.4f percent: generatorPktsSent %v, reflectorPktsRxed %v", tolerance, generatorPktsSent, reflectorPktsRxed)
-	}
-	if ((math.Abs(float64(reflectorPktsSent)-float64(generatorPktsRxed)))/(float64(reflectorPktsSent)+float64(generatorPktsRxed)+tolerance))*200.00 > tolerance {
-		t.Errorf("The difference between packets received count at Generator and packets sent count at Reflector is greater than %0.4f percent: reflectorPktsSent %v, generatorPktsRxed %v", tolerance, reflectorPktsSent, generatorPktsRxed)
+	if deviations.PLQReflectorStatsUnsupported(dut1) {
+		if (math.Abs(float64(generatorPktsSent)-float64(generatorPktsRxed))/float64(generatorPktsSent))*100.00 > tolerance {
```
Comment (LimeHat) [inline feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:443]: Why the change in math for the generator?

Based on the description, only the reflector check should be under the deviation.

**Review by LimeHat — CHANGES_REQUESTED** [review]

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go (hunk 1 above)
Comment (arvbaska1) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:443]: sorry i didn't understand , the existing checks are not changed they are moved to the else condition.
The existing checks are wrt to Generator sent - Reflector rcvd & Reflector sent - Generator rcvd,
so this wont work if reflector stats are unsupported, hence generator specific check is added under the deviation.

Comment (LimeHat) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:443]: Ah, I missed that part, sorry. Got it, thx.

**Comment by AmrNJ** [comment]
@arvbaska1 Please can you resolve conflicts?

**Comment by arvbaska1** [comment]
> @arvbaska1 Please can you resolve conflicts?

@AmrNJ conflicts are resolved now, could you please check and merge this

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go [hunk 2]
```diff
@@ -265,7 +282,7 @@ func TestLinkQualification(t *testing.T) {
 				EndpointType: &plqpb.QualificationConfiguration_PacketGenerator{
 					PacketGenerator: &plqpb.PacketGeneratorConfiguration{
 						PacketRate: uint64(138888),
```
Comment (dplore) [inline feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:284]:
```suggestion
						PacketRate: uint64(maxPPS),
```

**Review by dplore — COMMENTED** [review]
one suggestion, but otherwise looks good to me

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go (hunk 2 above)
Comment (arvbaska1) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:284]: 138888 PPS is used in base test https://github.com/openconfig/featureprofiles/blob/main/feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go#L267 right now.

Also not sure using MaxPPS will work with 100G link, even with below packet parameters it will be more than the bandwidth of a 100G link ?

40000000(PPS) *512 (Bytes) = 20480000000 Bps => 163.84 Gbps

**Comment by arvbaska1** [comment]
> revised comments.
> 
> Can you also fix the static check failure in this file? I know you didn't introduce it, but it is an easy fix.

okay, sure will change the function heading comment

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go [hunk 3]
```diff
@@ -44,6 +44,11 @@ func TestMain(m *testing.M) {
 //    https://github.com/fullstorydev/grpcurl
 //
 
+var (
+	maxMTU = uint64(8184)
+	maxPPS = uint64(1e8)
```
[the remark this answers is not in the document]
Comment (arvbaska1) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:49]: but this corresponding to the Generator capabilities here , so hence name as maxPPS and maxMTU.

https://github.com/openconfig/featureprofiles/blob/main/feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go#L92

https://github.com/openconfig/featureprofiles/blob/main/feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go#L100

[the remark this answers is not in the document]
Comment (arvbaska1) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:49]: But yes the value that is verified in Testcapabilities should be = > maxPPS/maxMTU value, so i can rename the variable as suggested above

**Comment by dplore** [comment]
Thanks for the updates, it makes more sense to me now.  Adding @nraghavendran to approve if the reduced PPS and MTU is an acceptable deviation to use.

**Review by dplore — APPROVED** [review]
LGTM.
